File: TorNetworkDefender.java
public class torNetworkDefense {

    /** Current State of Tor Network [Stage 1]; ideal attacker strategy - Regarding ASes: We operate on the Assumption (cited in the Paper) that 30% of autonomous systems are controlled by 6 AS
     * providers. Assuming a uniform distribution among these; that translates into each controlling ~5% of nodes. Also, these providers control
     * more bandwidth; henceforth we assume Tor gives them a degree of preferential treatment.
     *
     * 1. Separate pools for entry / exit nodes (~3% gain)
     * 2. First and Last Bridge outside of Tor cannot be from the same AS // ~2.7% additional gain along with (3), assumed a joint probabilistic way
     * 3. Asymmetric Routing of ASes eliminated / reduced // ~2.7% additional gain along with (2), assumed a joint probabilistic way
     * 4. Bandwidth requirement eliminated // ~4.5% gain in security
     * We Assume asymmetric routing within Tor - can we tackle this?
     */
    public static void main(String[] args) {
        int numberOfNodes = 6500;
        double exitNodes = 13;                   //this is a percentage value, i.e. exitNodes=5 means 5% of the nodes are exit eligible
        double maliciousExitNodes = 50;          //this is a percentage value, i.e. maliciousExitNodes=50 means 50% of exitNodes are malicious
        double maliciousNonExitNodes = 3;       //this is a percentage value, i.e. maliciousNonExitNodes=50 means 50% of nonExitNodes are malicious
        double repeats = 1000000;                //how many times we want to repeat the experiment
        double bandwidthThreshold = 1;           //Bandwidth threshold (from 1-100), here eliminated due to improvements
        double autonomousSystemsControl = 5;     //percent of 'external nodes' autonomous systems control (6 ASes control ~30%, assume uniform distr.; ~5% each)
        int numberOfASes = 6;                    // number of autonomous systems
        nodeSelector(numberOfNodes, exitNodes, maliciousExitNodes, maliciousNonExitNodes, repeats, bandwidthThreshold, autonomousSystemsControl, numberOfASes); //all nodes are within the Tor System
    }

    /**returns an array of Nodes with benevolent, malicious and exit nodes
     * 0 = benevolent node, 1 = exit node & benevolent node, 2 = malicious node & exit node, 3 = non-exit & malicious nodes
     * also assigns
     * These are nodes within the Tor System only
     */
    public static int[] createNodes(int numberOfNodes, double exitNodes, double maliciousExitNodes, double maliciousNonExitNodes){
        int[] nodes = new int[numberOfNodes];

        for (int i = 0; i < numberOfNodes; i++) nodes[i]=0;      // sets all nodes to be "benevolent" & non exit

        for (int i = 0; i < numberOfNodes ; i++) {             // randomly selects exit eligible and benevolent nodes
            int probability = StdRandom.uniform(100);
            if(probability<=exitNodes && nodes[i]!=1 && nodes[i]!=2) nodes[i]=1;
        }

        for (int i = 0; i < numberOfNodes; i++) {            // out of the exit eligible nodes selects a certain percentage to be malicious
            int probability = StdRandom.uniform(100);
            if(probability<=maliciousExitNodes && nodes[i]==1) nodes[i]=2;
        }

        for (int i = 0; i < numberOfNodes; i++) {            // out of the exit eligible nodes selects a certain percentage to be malicious
            int probability = StdRandom.uniform(100);
            if(probability<=maliciousNonExitNodes && nodes[i]==0) nodes[i]=3;
        }

        return nodes;
    }

    /**
     * Assigns a bandwidth value to all exit eligible nodes within the Tor Network
     */
    public static int[] assignBandwidthToNodes(int[] nodes, int numberOfNodes){
        int[] bandwidthOfNode = new int[numberOfNodes];
        for (int i = 0; i < numberOfNodes; i++) {
            if(nodes[i]==0) bandwidthOfNode[i]=0;                   // the bandwidth of non-exit eligible nodes is irrelevant for selection

            if(nodes[i]==1) bandwidthOfNode[i] = StdRandom.uniform(70); // non-malicious exit eligible nodes are uniformly distributed

            if(nodes[i]==2) bandwidthOfNode[i] = StdRandom.uniform(70) + 30; // the +30 is to show the preferences of attackers to occupy higher bandwidth nodes
        }
        return bandwidthOfNode;
    }

    /** Selects entry % exit nodes, and checks if both are malicious. Prints out how many times (of #of repeats) are both
     * selected nodes malicious, represents it as a percentage. It applies the bandwidth cutoff to Exit nodes; and also takes into account
     * ASes
     */
    public static void nodeSelector(int numberOfNodes, double exitNodes, double maliciousExitNodes, double maliciousNonExitNodes, double repeats, double bandwidthThreshold, double autonomousSystemsControl, int numberOfASes){
        int[] nodes = createNodes(numberOfNodes, exitNodes, maliciousExitNodes, maliciousNonExitNodes);
        int[] bandwidthOfNode = assignBandwidthToNodes(nodes, numberOfNodes);
        double caught = 0;

        for (int i = 0; i < repeats; i++) {
            int entryChoice = StdRandom.uniform(6500);                 //from individual to destination
            int exitChoice = StdRandom.uniform(6500);

            int backwardsEntryChoice = StdRandom.uniform(6500);                 //from destination to individual, backwards
            int backwardsExitChoice = StdRandom.uniform(6500);

            if (entryChoice == exitChoice || backwardsExitChoice == backwardsEntryChoice
                    || nodes[entryChoice]==2 || nodes[entryChoice]==1 ||
                    nodes[exitChoice]==0 || nodes[exitChoice]==3 || nodes[backwardsEntryChoice]==2 || nodes[backwardsEntryChoice]==1 ||
                    nodes[backwardsExitChoice]==0 || nodes[backwardsExitChoice]==3) {
                i--;
               // StdOut.println("Repeating"+i);
                continue;

            }

            if(bandwidthOfNode[exitChoice]<bandwidthThreshold || bandwidthOfNode[backwardsExitChoice]<bandwidthThreshold) {     //bandwidth condition, only applied to exit node
                i--;
                continue;
            }

           // if(nodes[exitChoice]<1 || nodes[backwardsExitChoice]<1) i--;                        //exit node condition (must be exit eligible)

            boolean autonomousCatch = autonomousSystemCatch(autonomousSystemsControl, numberOfASes);
            if((nodes[entryChoice]==3 && nodes[exitChoice]==2) || (nodes[backwardsEntryChoice] == 3 && nodes[backwardsExitChoice] == 2) ||
                    autonomousCatch) caught++;        // if both entry & exit are malicious
           // StdOut.println("Iteration" );
        }

        double probabilityOfGettingCaught = 100*caught/repeats;
        StdOut.print("The attacker succeeds " + probabilityOfGettingCaught + " % of the time within the Tor System.");
    }

    /**
     * Autonomous Systems; we return a boolean if ASes catch us (true); and false if they do not
     */
    public static boolean autonomousSystemCatch(double autonomousSystemsControl, int numberOfASes){
        for (int i = 0; i < numberOfASes; i++) {                             //repeat for each AS
            for (int j = 0; j < 2; j++) {                                   //send in both directions
                int beginning = StdRandom.uniform(100);
                int end = StdRandom.uniform(100);
                int weGetAway = StdRandom.uniform(100);
                if(weGetAway>80) return false;                        // represents steps 2, 3 in "improvement" probabilistically, notice that one should think of the complement probabilities here
                if(end==beginning){
                    j--;
                    continue;
                }

                if(end<=autonomousSystemsControl && beginning<=autonomousSystemsControl) return true;
            }
        }
        return false;
    }

}

File: OfficialResultsCode_C/TorNetworkDefender.c
#include <stdio.h>
#include <stdlib.h> 
#include <time.h> 
#include <math.h>

#define NODES 6500
#define EXIT 13			//percentage of nodes that are exit eligible 
#define MALICIOUS 50 	//percentage of nodes that are malicious
#define MALICIOUSE 3	//percentage of malicious NON EXIT nodes 
#define REPEATS 1000000 //number of times to repeat experiment
#define BANDWIDTHTH 1  //lower bandwidth threshhold
#define ASC 5 			//percentage of nodes controlled autonomously
#define NASC 6 			//number of autonomous systems

struct node
{
	int status;
	int bandwidth;
};

struct node network[NODES]; //work array 

//generates random # within range
int rand2(int max)
{
   return rand() / (RAND_MAX / (max + 1) + 1);
}

//init node status and assign bandwidth to it
//need to fix this 
void initNodes()
{ 
	for(int i = 0; i < NODES; i++)
	{
		network[i].status = 0;
		network[i].bandwidth = 0;

		int seed = rand2(100);
		if(seed <= EXIT)
		{
			seed = rand2(100);
			if(seed <= MALICIOUS)
			{
				network[i].status = 2;
				network[i].bandwidth = rand2(70) + 30;
			}
			else
			{
				network[i].status = 1;
				network[i].bandwidth = rand2(70);
			}
		}
		else 
		{
			seed = rand2(100);
			if(seed <= MALICIOUSE) network[i].status = 3;
		} 
	}
}

int asCatch()
{
	for(int i = 0; i < NASC; i++)
	{
		for(int j = 0; j < 2; j++)
		{
			int beginning = rand2(100);
			int end = rand2(100);
			int weGetAway = rand2(100);

			if(weGetAway > 80) return 0; //what is 80 here?

			if(end == beginning)
			{
				j--;
				continue;
			}

			if(end <= ASC && beginning <= ASC) return 1;
		}
	}

	return 0;
}

double selectNodes()
{
	double caught = 0.0;

	for(int i = 0; i < REPEATS; i++)
	{
		int entryChoice = rand2(NODES); //need to find threadsafe uniform rand2 function for C
		int exitChoice = rand2(NODES);

		int backwardsEntryChoice = rand2(NODES);
		int backwardsExitChoice = rand2(NODES);

		if(entryChoice == exitChoice || backwardsEntryChoice == backwardsExitChoice ||
		   network[entryChoice].status == 2 || network[entryChoice].status == 1 ||
		   network[exitChoice].status == 0 || network[exitChoice].status == 3 ||
		   network[backwardsEntryChoice].status == 2 || network[backwardsEntryChoice].status == 1 ||
		   network[backwardsExitChoice].status == 0 || network[backwardsExitChoice].status == 3)
		{
			i--;
			continue;
		}

		if(network[exitChoice].bandwidth < BANDWIDTHTH || network[backwardsExitChoice].bandwidth < BANDWIDTHTH)
		{
			i--;
			continue;
		}

		int AS = asCatch();

		if((network[entryChoice].status == 3 && network[exitChoice].status == 2) || 
		   (network[backwardsEntryChoice].status == 3 && network[backwardsExitChoice].status == 2) || AS)
		{
			caught += 1.0;
		}
	}

	double caughtProbability = 100.0 * caught / (double)REPEATS;
	return caughtProbability;
}

int main(int argc, char* argv[])
{
	srand(time(NULL)); //seed RNG 

	initNodes();
	double result = selectNodes();

	printf("Attacker success rate is %lf\n", result); //possibly write to CSV? 
}